### PR TITLE
Docs/jbdirectory

### PR DESCRIPTION
**Pull request by mejango**
- `setTerminalsOf` instead of `addTerminalsOf` and `removeTerminalOf`.
- `setIsAllowedToSetFirstController` instead of `addToSetControllerAllowList` and `removeFromSetControllerAllowList`
- contract that are on the controller setting whitelist can only set a project's controller if it currently has no controller set. Controller being set no longer needs to also be on the allow list.
- remove checks for if setting controller to a zero address or the current address.
- doc updates.

needs tests. It'll be a net reduction of tests. @drgorillamd

### CODE REVIEW

File: contracts/JBDirectory.sol [hunk 1]
```diff
@@ -350,9 +304,27 @@ contract JBDirectory is IJBDirectory, JBOperatable, Ownable {
...
+  /** 
+    @notice
+    Check if the provided terminal array contains the provided terminal.
+
+    @param _terminals The terminals to look through.
+    @param _terminal The terminal to check for.
+
+    @return Whether or not the `_terminals` includes the `_terminal`.
+  */
+  function _contains(IJBPaymentTerminal[] calldata _terminals, IJBPaymentTerminal _terminal)
```
Comment (drgorillamd) [inline contracts/JBDirectory.sol:322]: me see calldata, me like

File: contracts/JBDirectory.sol [hunk 2]
```diff
@@ -198,63 +200,39 @@ contract JBDirectory is IJBDirectory, JBOperatable, Ownable {
...
-    delete _terminalsOf[_projectId];
+    _terminalsOf[_projectId] = _terminals;
 
-    // Repopulate the stored terminals for the project, omitting the one being deleted.
-    for (uint256 _i; _i < _terminals.length; _i++)
-      // Don't include the terminal being deleted.
-      if (_terminals[_i] != _terminal) _terminalsOf[_projectId].push(_terminals[_i]);
+    // If one of the old terminals was set as a primary terminal but is not included in the new terminals, remove it from being a primary terminal.
+    for (uint256 _i; _i < _oldTerminals.length; _i++)
+      // If the terminal that is being removed is the primary terminal for the token, delete it from being primary terminal.
```
Comment (drgorillamd) [inline contracts/JBDirectory.sol:229]: deprecated comment (cf supra)?

File: contracts/JBDirectory.sol [hunk 3]
```diff
@@ -198,63 +200,39 @@ contract JBDirectory is IJBDirectory, JBOperatable, Ownable {
...
+      if (
+        _primaryTerminalOf[_projectId][_terminals[_i].token()] == _oldTerminals[_i] &&
+        !_contains(_terminals, _terminals[_i])
+      ) delete _primaryTerminalOf[_projectId][_oldTerminals[_i].token()];
```
Comment (drgorillamd) [inline contracts/JBDirectory.sol:233]: break;

File: contracts/JBDirectory.sol (hunk 2 above)
Comment (mejango) [reply contracts/JBDirectory.sol]: deleted the second comment line

File: contracts/JBDirectory.sol (hunk 1 above)
Comment (mejango) [reply contracts/JBDirectory.sol:321]: slurp

File: contracts/JBDirectory.sol [hunk 4]
```diff
@@ -198,63 +200,39 @@ contract JBDirectory is IJBDirectory, JBOperatable, Ownable {
...
-    override
-    requirePermission(projects.ownerOf(_projectId), _projectId, JBOperations.REMOVE_TERMINAL)
   {
     // Get a reference to the terminals of the project.
-    IJBPaymentTerminal[] memory _terminals = _terminalsOf[_projectId];
+    IJBPaymentTerminal[] memory _oldTerminals = _terminalsOf[_projectId];
 
     // Delete the stored terminals for the project.
-    delete _terminalsOf[_projectId];
+    _terminalsOf[_projectId] = _terminals;
```
[the remark this answers is not in the document]
Comment (mejango) [reply contracts/JBDirectory.sol:225]: oo. true. doesn't hurt but not ideal.

File: contracts/JBDirectory.sol [hunk 5]
```diff
@@ -198,63 +200,39 @@ contract JBDirectory is IJBDirectory, JBOperatable, Ownable {
...
-    // Repopulate the stored terminals for the project, omitting the one being deleted.
-    for (uint256 _i; _i < _terminals.length; _i++)
-      // Don't include the terminal being deleted.
-      if (_terminals[_i] != _terminal) _terminalsOf[_projectId].push(_terminals[_i]);
+    // If one of the old terminals was set as a primary terminal but is not included in the new terminals, remove it from being a primary terminal.
+    for (uint256 _i; _i < _oldTerminals.length; _i++)
+      // If the terminal that is being removed is the primary terminal for the token, delete it from being primary terminal.
+      if (
+        _primaryTerminalOf[_projectId][_terminals[_i].token()] == _oldTerminals[_i] &&
+        !_contains(_terminals, _terminals[_i])
```
[the remark this answers is not in the document]
Comment (mejango) [reply contracts/JBDirectory.sol:231]: oo. true. should be _oldTerminals[_i] not _terminals[_i]

File: contracts/JBDirectory.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (mejango) [reply contracts/JBDirectory.sol:223]: added a loop to check for this

**Comment by mejango** [comment]
take another look @drgorillamd. we back at 100 cov. down to merge when it looks good to you. 
CI tests seem to be working fine.

File: test/jb_directory/set_terminals_of.test.js [hunk 6]
```diff
@@ -72,7 +73,7 @@ describe('JBDirectory::setTerminalsOf(...)', function () {
 
     await expect(jbDirectory.connect(projectOwner).setTerminalsOf(PROJECT_ID, terminals))
       .to.emit(jbDirectory, 'SetTerminals')
-      //.withArgs(PROJECT_ID, terminals, projectOwner.address); -->  AssertionError: expected undefined to equal (...) ?!
+    //.withArgs(PROJECT_ID, terminals, projectOwner.address); -->  AssertionError: expected undefined to equal (...) ?!
```
Comment (drgorillamd) [inline test/jb_directory/set_terminals_of.test.js:76]: Gave it another try, still don't know how to solve this (event is emitted, as tested, tho)

**Review by drgorillamd — APPROVED** [review]

File: test/jb_directory/set_terminals_of.test.js (hunk 6 above)
Comment (mejango) [reply test/jb_directory/set_terminals_of.test.js:76]: can we get this with a forge test?

Comment (mejango) [reply test/jb_directory/set_terminals_of.test.js]: fixed by removing `indexed` from terminals arg to the event

**Pull request merged**